geoh5py/groups/property_group.py

```diff
@@ -56,67 +71,180 @@ class PropertyGroup:
+            value = getattr(DataAssociationEnum, value.upper())
 
-        map_attributes(self, **kwargs)
+        if not isinstance(value, DataAssociationEnum):
+            raise TypeError(f"Association must be one of type {DataAssociationEnum}")
 
-        self.parent.workspace.register(self)
+        return value
 
-    def add_properties(self, data: Data | list[Data | uuid.UUID] | uuid.UUID):
+    def _validate_data(self, data: Data | UUID | str) -> UUID:
         """
-        Remove data from the properties.
+        Verify that the data is in the parent and has the same association as the group.
+
+        :param data: The data to verify.
+            It can be the name, the uuid or the data itself.
+
+        :return: The uuid of the data.
+        """
+        data = self.parent.reference_to_data(data)
+
+        if self.association != data.association:
+            raise ValueError(
+                f"Data '{data.name}' association ({data.association}) "
+                f"does not match group association ({self.association})."
+            )
+
+        return data.uid
+
+    @staticmethod
+    def _validate_group_type(value: str | GroupTypeEnum) -> GroupTypeEnum:
+        """
+        Verify that the group type is a valid GroupTypeEnum.
+
+        :param value: The group type to validate.
+
+        :return: The validated group type.
+        """
+        if isinstance(value, str):
+            try:
+                value = GroupTypeEnum(value)
+            except ValueError as error:
+                raise ValueError(
+                    f"'Property group type' must be one of "
+                    f"{', '.join(GroupTypeEnum.__members__)}. Provided {value}"
+                ) from error
+
+        if not isinstance(value, GroupTypeEnum):
+            raise TypeError(
+                f"'Property group type' must be of type {GroupTypeEnum}, "
+                f"provided {type(value)}"
+            )
+
+        return value
+
+    @staticmethod
+    def _validate_parent(parent: ObjectBase) -> ObjectBase:
+        """
+        Verify that the parent is valid.
+
+        :param parent: The parent Object to validate.
+
+        :return: The parent Object.
+        """
+        # define the parent
+        if not hasattr(parent, "_property_groups"):
+            raise TypeError(f"Parent {parent} must have a 'property_groups' attribute")
+        return parent
+
+    def _validate_properties(self, data_list: list[Data] | None) -> list[UUID] | None:
         """
-        if isinstance(data, (Data, uuid.UUID)):
+        Validate the properties list.
+
+        :param data_list: List of Data entities to validate.
+
+        :return: List of unique identifiers for the Data entities.
+        """
+        if data_list is None:
+            return None
+
+        if not isinstance(data_list, Iterable):
+            raise TypeError(
+                f"Properties must be an iterable of Data. Provided {data_list}"
+            )
```

Comment:
I think we can remove. This would already be covered by _initialize_properties. I wonder if we could not remove this method altogether and just to the list conversion directly in the init?

Reply:
right, it's not pertinent anymore